### ISSUE TITLE
Consumption Reporting - Support for location reporting

### ISSUE THREAD

**Issue by dsilhavy** [5G-MAG/rt-5gms-media-stream-handler#53]
## Feature description
As part of the ConsumptionReportingUnit type as defined in 26.512 Table 11.3.3.2 the `locations` property defines a time-ordered list of UE location(s) where the media was consumed during the period of this consumption reporting unit.

The `TypedLocation` type for this property is defined in 6.4.3.6

## Implementation
It is to be identified how to derive this information via an Android API.

**Comment by dsilhavy** [comment]
The TelephonyManager API possibly provides the information we need: https://developer.android.com/reference/android/telephony/TelephonyManager#getAllCellInfo()

There is a specific`CellInfoLte.getCellIdentity()` method which we can possibly use for `CellIdentifierType` set to `ECGI`: https://developer.android.com/reference/android/telephony/CellInfoLte#getCellIdentity()

I am getting some warnings in the code though that `getCellIdendity` requires a minimum API level of 30. This exceeds the current minimum API level and the capabilities of my test device which supports API level 29.

**Comment by dsilhavy** [comment]
Addition: Android API level 30 corresponds to Android 11 as documented here: https://apilevels.com/

**Comment by dsilhavy** [comment]
The following basic function to populate the `locations` property was implemented:

````kotlin
    private fun getLocations(): ArrayList<TypedLocation> {
        val locations = ArrayList<TypedLocation>()
        val telephonyManager =
            context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
        try {
            val cellInfoList: List<CellInfo> = telephonyManager.allCellInfo

            for (cellInfo in cellInfoList) {
                if (cellInfo.isRegistered) {
                    val cellIdentity = cellInfo.cellIdentity
                    val location = cellIdentity.operatorAlphaLong.toString()
                    locations.add(TypedLocation(CellIdentifierType.CGI, location))
                }
            }
        } catch (e: SecurityException) {
            return ArrayList()
        }

        return locations
    }
````


In my tests ` telephonyManager.allCellInfo` always returned an empty array even if the device is connected to a mobile network. This needs to be evaluated further. It might be a security restriction on the platform/device and this information is possibly only available to system processes.

**Comment by rjb1000** [comment]
I think the value of the `CellIdentifierType` enumeration will depend on whether the `CellInfo` is a `CellInfoGsm`, `CellInfoLte`, `CellInfoNr`, etc.

**Comment by rjb1000** [comment]
Does `operatorAlphaLong.toString()` really yield the raw Cell Global Identity value of a `CellInfo`?

**Comment by dsilhavy** [comment]
Thanks @rjb1000 

> I think the value of the `CellIdentifierType` enumeration will depend on whether the `CellInfo` is a `CellInfoGsm`, `CellInfoLte`, `CellInfoNr`, etc.

Yes I agree, once I figured out how to get actual cell info I will adjust the implementation accordingly.

> Does `operatorAlphaLong.toString()` really yield the raw Cell Global Identity value of a `CellInfo`?

Need to verify this, when I quickly scanned the API docs this is what I found. But I assume that there are other methods to check based on the type of the cell like `CellInfoLte` and `CellInfoGsm` e.g. https://developer.android.com/reference/android/telephony/CellIdentityGsm#getCid()  .

**Comment by dsilhavy** [comment]
I checked again how to derive useful information about the cells:

I wrote multiple instrumental tests in the MediaStreamHandler to check if I can derive any useful values with the following functions. The reason to implement this with instrumental tests was to be able to set breakpoints in the code as the MediaStreamHandler is only a library.
  * https://developer.android.com/reference/android/telephony/TelephonyManager#requestCellInfoUpdate(java.util.concurrent.Executor,%20android.telephony.TelephonyManager.CellInfoCallback)  Note: With Android Q (10) the correct way to request cell information seems to be this function.
  * https://developer.android.com/reference/android/telephony/TelephonyManager#getAllCellInfo()
  * https://developer.android.com/reference/android/telephony/TelephonyManager#getCellLocation()

In all cases, the list of cells was empty.

Next, I implemented a small standalone app that runs in the foreground and explicitly asks the user for permission to access location information. This seems to to do the trick as I am getting a list of cells now in my callback function, for example:

````
CellInfoLte:{mRegistered=YES mTimeStamp=12819893611791ns mCellConnectionStatus=0 CellIdentityLte:{ mCi=<MASKED> mPci=125 mTac=<MASKED> mEarfcn=1600 mBandwidth=2147483647 mMcc=262 mMnc=03 mAlphaLong=o2 - de mAlphaShort=o2 - de} CellSignalStrengthLte: rssi=2147483647 rsrp=-109 rsrq=-16 rssnr=-3 cqi=2147483647 ta=4 level=1 android.telephony.CellConfigLte :{ isEndcAvailable = false }}

CellInfoLte:{mRegistered=NO mTimeStamp=12819893611791ns mCellConnectionStatus=0 CellIdentityLte:{ mCi=<MASKED> mPci=141 mTac=<MASKED> mEarfcn=1600 mBandwidth=2147483647 mMcc=null mMnc=null mAlphaLong= mAlphaShort=} CellSignalStrengthLte: rssi=2147483647 rsrp=-112 rsrq=-19 rssnr=0 cqi=0 ta=0 level=1 android.telephony.CellConfigLte :{ isEndcAvailable = false }}
````

The next step is to check which information about a cell is available and how this can be mapped to required information as specified in 26512. For instance, `cellInfoList[0].cellIdentity` returns:

````
CellIdentityLte:{ mCi=<MASKED> mPci=125 mTac=<MASKED> mEarfcn=1600 mBandwidth=2147483647 mMcc=262 mMnc=03 mAlphaLong=o2 - de mAlphaShort=o2 - de}
````

**Comment by dsilhavy** [comment]
Based on the discussion in the last call, we identified the following logic to derive the `CGI`, ' ECGI` and `NCGI`:

* CGI = MCC + MNC + LAC + CI
* ECGI = MCC + MNC + ECI
* NCGI = MCC + MNC + NCI

For details check TS23003

**Comment by dsilhavy** [comment]
> Based on the discussion in the last call, we identified the following logic to derive the `CGI`, `ECGI` and `NCGI`:
> 
> * CGI = MCC + MNC + LAC + CI
> * ECGI = MCC + MNC + ECI
> * NCGI = MCC + MNC + NCI
> 
> For details check TS23003

Implemented accordingly, as an example:

````json
{
  "mediaPlayerEntry": "https://dash.akamaized.net/envivio/EnvivioDash3/manifest.mpd",
  "reportingClientId": "918af625-31d1-473c-af0b-8da283b6de2c",
  "consumptionReportingUnits": [
    {
      "mediaConsumed": "v4_258",
      "mediaEndpointAddress": {
        "domainName": null,
        "ipv4Addr": "192.168.178.40",
        "ipv6Addr": null,
        "portNumber": 80
      },
      "startTime": "2023-11-06T16:38:33Z",
      "duration": 41,
      "locations": [
        {
          "locationIdentifierType": "ECGI",
          "location": "262036460715"
        }
      ],
      "mimeType": "audio/mp4",
      "finished": false
    },
    {
      "mediaConsumed": "v7_257",
      "mediaEndpointAddress": {
        "domainName": null,
        "ipv4Addr": "192.168.178.40",
        "ipv6Addr": null,
        "portNumber": 80
      },
      "startTime": "2023-11-06T16:38:45Z",
      "duration": 29,
      "locations": [
        {
          "locationIdentifierType": "ECGI",
          "location": "262036460715"
        }
      ],
      "mimeType": "video/mp4",
      "finished": false
    }
  ]
}
````

Some comments:
* I could only test the reporting for `ECGI` but `CGI` and `NCGI` should work as well
* An entry to `locations` is only added if `cellInfo.cellConnectionStatus == CellInfo.CONNECTION_PRIMARY_SERVING || cellInfo.cellConnectionStatus == CellInfo.CONNECTION_SECONDARY_SERVING`
* I rewrote some parts in the code to improve the structure for consumption reporting

I still need to implement the logic to detect location changes and trigger a consumption report. Afterwards I will open the pull request with all the changes above.

**Comment by dsilhavy** [comment]
Implemented in #58 as described above, including sending a consumption report once the cellInfo changes.
